File: web/lib/solar.js
// This Source Code Form is subject to the terms of the Mozilla Public
// License, v. 2.0. If a copy of the MPL was not distributed with this
// file, You can obtain one at https://mozilla.org/MPL/2.0/.

// Where the sun is, which is what draws the greyline.
//
// The terminator is the great circle 90 degrees from the subsolar point, so
// once you know where the sun is directly overhead the rest is geometry. This
// is the standard low-precision solar position algorithm, good to about a
// hundredth of a degree -- far better than a few hundred pixels of globe can
// show, and it runs in the browser with no data and no network.

const DEG = Math.PI / 180;
const TAU = Math.PI * 2;

/** Julian centuries-free day number since J2000.0. */
function daysSinceJ2000(date) {
  return date.getTime() / 86400000 + 2440587.5 - 2451545.0;
}

/**
 * The point on Earth where the sun is directly overhead, right now.
 * @returns {{lat: number, lon: number}} degrees
 */
export function subsolarPoint(date = new Date()) {
  const n = daysSinceJ2000(date);

  // Mean longitude and mean anomaly of the sun.
  const meanLon = (280.46 + 0.9856474 * n) * DEG;
  const meanAnomaly = (357.528 + 0.9856003 * n) * DEG;

  // Ecliptic longitude, with the two largest periodic corrections.
  const eclipticLon =
    meanLon + 1.915 * DEG * Math.sin(meanAnomaly) + 0.02 * DEG * Math.sin(2 * meanAnomaly);

  const obliquity = (23.439 - 0.0000004 * n) * DEG;

  const declination = Math.asin(Math.sin(obliquity) * Math.sin(eclipticLon));
  const rightAscension = Math.atan2(
    Math.cos(obliquity) * Math.sin(eclipticLon),
    Math.cos(eclipticLon),
  );

  // Greenwich mean sidereal time tells us which meridian faces the sun.
  const gmstHours = (18.697374558 + 24.06570982441908 * n) % 24;
  const gmst = ((gmstHours + 24) % 24) * 15 * DEG;

  let lon = (rightAscension - gmst) / DEG;
  lon = ((((lon + 180) % 360) + 360) % 360) - 180;

  return { lat: declination / DEG, lon };
}

/**
 * Points on the terminator: the circle 90 degrees from the subsolar point.
 * @returns {Array<{lat: number, lon: number}>}
 */
export function terminatorRing(subsolar, steps = 360) {
  const lat1 = subsolar.lat * DEG;
  const lon1 = subsolar.lon * DEG;
  const d = Math.PI / 2; // a quarter turn away, by definition

  const ring = [];
  for (let i = 0; i <= steps; i += 1) {
    const bearing = (i / steps) * TAU;
    const lat = Math.asin(
      Math.sin(lat1) * Math.cos(d) + Math.cos(lat1) * Math.sin(d) * Math.cos(bearing),
    );
    const lon =
      lon1 +
      Math.atan2(
        Math.sin(bearing) * Math.sin(d) * Math.cos(lat1),
        Math.cos(d) - Math.sin(lat1) * Math.sin(lat),
      );
    ring.push({ lat: lat / DEG, lon: (((lon / DEG + 540) % 360) - 180) });
  }
  return ring;
}

/** Solar elevation at a point, in degrees. Negative is night. */
export function solarElevation(lat, lon, subsolar) {
  const a = lat * DEG;
  const b = subsolar.lat * DEG;
  const dl = (lon - subsolar.lon) * DEG;
  const cosZenith = Math.sin(a) * Math.sin(b) + Math.cos(a) * Math.cos(b) * Math.cos(dl);
  return Math.asin(Math.max(-1, Math.min(1, cosZenith))) / DEG;
}

/** True when a point is in darkness. */
export function isNight(lat, lon, subsolar) {
  return solarElevation(lat, lon, subsolar) < 0;
}

/**
 * How close a point is to the greyline, as a 0..1 weight.
 *
 * The band either side of sunrise and sunset is where the low bands do
 * remarkable things, so it is worth showing as more than a hard edge.
 */
export function greylineWeight(lat, lon, subsolar, halfWidthDegrees = 6) {
  const elevation = Math.abs(solarElevation(lat, lon, subsolar));
  return elevation >= halfWidthDegrees ? 0 : 1 - elevation / halfWidthDegrees;
}

File: src/hammunition_hill/solar.py
# This Source Code Form is subject to the terms of the Mozilla Public
# License, v. 2.0. If a copy of the MPL was not distributed with this
# file, You can obtain one at https://mozilla.org/MPL/2.0/.

"""Where the sun is, in Python.

The browser already has this in ``web/lib/solar.js`` -- it draws the greyline.
The collector needs it too, because the propagation model is computed there and
its single most important input is *how high the sun is above the operator's
own station*, not what hour it happens to be in Greenwich.

This is the standard low-precision solar position algorithm, good to about a
hundredth of a degree. That is far more than enough: the ionospheric models
downstream are empirical approximations whose error is measured in megahertz,
so refining the sun's position past this point would be polishing the one part
that was never the problem.

Deliberately duplicated rather than shared. The browser copy has no build step
and no imports by design; making both sides read one file would mean either
shipping a Python-to-JS build or serving the maths as data. Two ~40-line
implementations of a published algorithm, with tests that pin them to the same
reference values, is the cheaper trade.
"""

from __future__ import annotations

import math
from dataclasses import dataclass
from datetime import UTC, datetime

# Unix epoch as a Julian day number, minus the J2000.0 epoch.
_JD_UNIX_EPOCH = 2440587.5
_JD_J2000 = 2451545.0
_SECONDS_PER_DAY = 86400.0


@dataclass(frozen=True)
class Subsolar:
    """The point on Earth where the sun is directly overhead."""

    lat: float
    lon: float


def days_since_j2000(moment: datetime) -> float:
    """Days since J2000.0. Naive datetimes are assumed UTC, not local."""
    if moment.tzinfo is None:
        moment = moment.replace(tzinfo=UTC)
    return moment.timestamp() / _SECONDS_PER_DAY + _JD_UNIX_EPOCH - _JD_J2000


def subsolar_point(moment: datetime | None = None) -> Subsolar:
    """Where the sun is overhead right now."""
    n = days_since_j2000(moment or datetime.now(UTC))

    mean_lon = math.radians(280.46 + 0.9856474 * n)
    mean_anomaly = math.radians(357.528 + 0.9856003 * n)

    # Ecliptic longitude, with the two largest periodic corrections.
    ecliptic_lon = (
        mean_lon
        + math.radians(1.915) * math.sin(mean_anomaly)
        + math.radians(0.020) * math.sin(2 * mean_anomaly)
    )
    obliquity = math.radians(23.439 - 0.0000004 * n)

    declination = math.asin(math.sin(obliquity) * math.sin(ecliptic_lon))
    right_ascension = math.atan2(
        math.cos(obliquity) * math.sin(ecliptic_lon), math.cos(ecliptic_lon)
    )

    # Greenwich mean sidereal time says which meridian is facing the sun.
    gmst_hours = (18.697374558 + 24.06570982441908 * n) % 24
    gmst = math.radians(((gmst_hours + 24) % 24) * 15)

    lon = math.degrees(right_ascension - gmst)
    lon = ((lon + 180) % 360 + 360) % 360 - 180

    return Subsolar(lat=math.degrees(declination), lon=lon)


def solar_elevation(lat: float, lon: float, subsolar: Subsolar) -> float:
    """Sun's angle above the horizon at a point, in degrees. Negative is night."""
    a = math.radians(lat)
    b = math.radians(subsolar.lat)
    delta_lon = math.radians(lon - subsolar.lon)
    cos_zenith = math.sin(a) * math.sin(b) + math.cos(a) * math.cos(b) * math.cos(delta_lon)
    return math.degrees(math.asin(max(-1.0, min(1.0, cos_zenith))))


def solar_zenith(lat: float, lon: float, subsolar: Subsolar) -> float:
    """Angle from straight up, in degrees. 0 is overhead, 90 is the horizon.

    This is the quantity the ionospheric models actually want. Absorption
    scales with the cosine of it, which is why a solar-noon *proxy* built from
    the UTC hour is not good enough: it is exactly right on the Greenwich
    meridian and hours wrong everywhere else.
    """
    return 90.0 - solar_elevation(lat, lon, subsolar)


def is_night(lat: float, lon: float, subsolar: Subsolar) -> bool:
    return solar_elevation(lat, lon, subsolar) < 0.0
